src/DataGrid.tsx

```diff
@@ -514,21 +514,27 @@ function DataGrid<R, SR, K extends Key>(
     const { row, checked, isShiftClick } = args;
     const newSelectedRows = new Set(selectedRows);
     const rowKey = rowKeyGetter(row);
+    const previousRowIdx = lastSelectedRowIdx.current;
+    const rowIdx = rows.indexOf(row);
+    lastSelectedRowIdx.current = rowIdx;
+
     if (checked) {
       newSelectedRows.add(rowKey);
-      const previousRowIdx = lastSelectedRowIdx.current;
-      const rowIdx = rows.indexOf(row);
-      lastSelectedRowIdx.current = rowIdx;
-      if (isShiftClick && previousRowIdx !== -1 && previousRowIdx !== rowIdx) {
-        const step = sign(rowIdx - previousRowIdx);
-        for (let i = previousRowIdx + step; i !== rowIdx; i += step) {
-          const row = rows[i];
+    } else {
+      newSelectedRows.delete(rowKey);
+    }
+
+    if (isShiftClick && previousRowIdx !== -1 && previousRowIdx !== rowIdx) {
+      const step = sign(rowIdx - previousRowIdx);
+      for (let i = previousRowIdx + step; i !== rowIdx; i += step) {
+        const row = rows.at(i);
+        if (row === undefined) continue;
```

Comment:
Added this extra check in case the previous row does not exist. Is there a better way?

Reply:
If the previous row does not exist I'd assume we don't want to apply the shift+click behavior?
We could check `rows.length` no?

Reply:
You are right. I made that change